File: rxhive-components/rxhive-components-akkastream_2.12/src/main/kotlin/com/sksamuel/reactivehive/akkastream/ParquetSource.kt
package com.sksamuel.reactivehive.akkastream

import akka.stream.Attributes
import akka.stream.Outlet
import akka.stream.SourceShape
import akka.stream.stage.GraphStageLogic
import akka.stream.stage.GraphStageWithMaterializedValue
import akka.stream.stage.OutHandler
import com.sksamuel.reactivehive.Struct
import com.sksamuel.reactivehive.parquet.parquetReader
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import scala.Tuple2
import java.util.concurrent.CompletableFuture

class ParquetSource(val path: Path,
                    val conf: Configuration) : GraphStageWithMaterializedValue<SourceShape<Struct>, CompletableFuture<Int>>() {

  private val outlet: Outlet<Struct> = Outlet.create("ParquetSource.out")
  override fun shape(): SourceShape<Struct> = SourceShape.of(outlet)

  override fun createLogicAndMaterializedValue(inheritedAttributes: Attributes?): Tuple2<GraphStageLogic, CompletableFuture<Int>> {

    val future = CompletableFuture<Int>()

    val graph: GraphStageLogic = object : GraphStageLogic(shape()), OutHandler {

      private val reader = parquetReader(path, conf)
      private var count = 0

      init {
        setHandler(outlet, this)
      }

      override fun onPull() {
        val struct = reader.read()
        try {
          if (struct == null) {
            completeStage()
          } else {
            push(outlet, struct)
            count++
          }
        } catch (t: Throwable) {
          failStage(t)
        }
      }
    }

    return Tuple2(graph, future)
  }
}

File: centurion-components/rxhive-akkastream_2.12/src/main/scala/com/sksamuel/rxhive/akkastream/ParquetSource.scala
package com.sksamuel.rxhive.akkastream

import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, OutHandler}
import akka.stream.{Attributes, Outlet, SourceShape}
import com.sksamuel.rxhive.Struct
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path

import scala.concurrent.{Future, Promise}

class ParquetSource(val path: Path,
                    val conf: Configuration) extends GraphStageWithMaterializedValue[SourceShape[Struct], Future[Long]]() {

  private val outlet: Outlet[Struct] = Outlet.create("ParquetSource.out")
  override def shape(): SourceShape[Struct] = SourceShape.of(outlet)

  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Long]) = {

    val promise = Promise[Long]

    val graph: GraphStageLogic = new GraphStageLogic(shape()) with OutHandler {

      private val reader = com.sksamuel.rxhive.parquet.ParquetReaderKt.parquetReader(path, conf)
      private var count = 0L

      setHandler(outlet, this)

      override def onPull() {
        val struct = reader.read()
        try {
          if (struct == null) {
            completeStage()
            promise.trySuccess(count)
          } else {
            push(outlet, struct)
            count = count + 1
          }
        } catch {
          case t: Throwable =>
            failStage(t)
            promise.tryFailure(t)
        }
      }
    }

    Tuple2(graph, promise.future)
  }
}